From 0a5850f11cc9364fb7e574d1c27f0b1e3ee86816 Mon Sep 17 00:00:00 2001
From: Shadi Ammouri <shadi@marvell.com>
Date: Sun, 16 Dec 2012 13:47:32 +0200
Subject: [PATCH 6/7] ARM: armadaxp: changed the internal register base
 address

To upgrade the u-boot version of armdaxp to 2013_Q1.2, the corresponding
internal register base address needs to be changed in linux kernel.

The code is extracted from linux-3.2.y-2013_Q1.2, which
can be downloaded from:
https://extranet.marvell.com/extranet/dms/documents.do?groupID=4&subGroupID=55266

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 arch/arm/boot/compressed/head.S                |   14 ++++++++++++++
 arch/arm/mach-armadaxp/include/mach/armadaxp.h |   20 ++------------------
 2 files changed, 16 insertions(+), 18 deletions(-)

diff --git a/arch/arm/boot/compressed/head.S b/arch/arm/boot/compressed/head.S
index f2db248..3a00813 100644
--- a/arch/arm/boot/compressed/head.S
+++ b/arch/arm/boot/compressed/head.S
@@ -9,6 +9,9 @@
  * published by the Free Software Foundation.
  */
 #include <linux/linkage.h>
+#ifdef CONFIG_ARCH_ARMADA_XP
+#include <mach/armadaxp.h>
+#endif
 
 	.arch	armv7-a
 /*
@@ -120,6 +123,17 @@
 		.align
 		.arm				@ Always enter in ARM state
 start:
+#ifdef CONFIG_ARCH_ARMADA_XP
+		/* Update Internal Regs offset in case UBoot is configured
+		** to use a different base address.
+		*/
+		mrc p15, 4, r0, c15, c0, 0	@ Get the internal registers base address
+		lsl r0, r0, #13			@ the address is R-shifted, need to recover it
+		ldr r5, =0x20080
+		add r0, r0, r5
+		ldr r6, =INTER_REGS_PHYS_BASE
+		str r6, [r0]
+#endif
 		.type	start,#function
 		.rept	7
 		mov	r0, r0
diff --git a/arch/arm/mach-armadaxp/include/mach/armadaxp.h b/arch/arm/mach-armadaxp/include/mach/armadaxp.h
index c5bb6b3..8db1acf 100644
--- a/arch/arm/mach-armadaxp/include/mach/armadaxp.h
+++ b/arch/arm/mach-armadaxp/include/mach/armadaxp.h
@@ -98,15 +98,7 @@
 #define SPI_CS0_VIRT_BASE		0xFAA00000
 #define SPI_CS0_SIZE			_16M
 
-#ifdef CONFIG_MACH_ARMADA_XP_FPGA
- #define INTER_REGS_PHYS_BASE		0xF1000000
- /* Make sure that no other machines are compiled in */
- #if defined (CONFIG_MACH_ARMADA_XP_DB) || defined (CONFIG_MACH_ARMADA_XP_RDSRV)
- #error	"Conflicting Board Configuration!!"
- #endif
-#else
- #define INTER_REGS_PHYS_BASE		0xD0000000
-#endif
+#define INTER_REGS_PHYS_BASE		0xF1000000
 
 /*
  * Change INTER_REGS_BASE from 0xFBB00000 to 0xFBC00000 is mainly
@@ -208,15 +200,7 @@
 #define SPI_CS0_VIRT_BASE		0xFAB00000
 #define SPI_CS0_SIZE			_16M
 
-#ifdef CONFIG_MACH_ARMADA_XP_FPGA
- #define INTER_REGS_PHYS_BASE		0xF1000000
- /* Make sure that no other machines are compiled in */
- #if defined (CONFIG_MACH_ARMADA_XP_DB) || defined (CONFIG_MACH_ARMADA_XP_RDSRV)
- #error	"Conflicting Board Configuration!!"
- #endif
-#else
- #define INTER_REGS_PHYS_BASE		0xD0000000
-#endif
+#define INTER_REGS_PHYS_BASE		0xF1000000
 #define INTER_REGS_BASE			0xFBB00000
 
 #define PEX0_IO_PHYS_BASE		0xF1100000
-- 
1.7.5.4

